pkg/power/redfish/redfish.go

```diff
@@ -0,0 +1,255 @@
+	Password string `json:"password"`
+	Host     string `json:"host"`
+}
+
+type RedFishClient struct {
+	// systemEnergy is the system accumulated energy consumption in Joule
+	accessInfo RedfishAccessInfo
+	systems    []*RedfishSystemPowerResult
+	ticker     *time.Ticker
+	mutex      sync.Mutex
+}
+
+func NewRedfishClient(username, password, host string) *RedFishClient {
+	redfish := &RedFishClient{
+		accessInfo: RedfishAccessInfo{
+			Username: username,
+			Password: password,
+			Host:     host,
+		},
+	}
+	if redfish.IsPowerSupported() {
+		klog.Infof("starting redfish power collection\n")
+	}
+	return redfish
+}
+
+func (rf *RedFishClient) IsPowerSupported() bool {
+	system, err := getRedfishSystem(rf.accessInfo)
+
+	if err != nil {
+		klog.V(5).Infof("failed to get system info: %v\n", err)
+		return false
+	}
+
+	intervalInMin := 0
+	// iterate each "Members" in the system and get the power info
+	for _, member := range system.Members {
+		// split the OdataID by delimiter "/" and get the system ID
+		split := strings.Split(member.OdataID, "/")
+		if len(split) < 2 {
+			continue
+		}
+		id := split[len(split)-1]
+		res := RedfishSystemPowerResult{}
+		power, err := getRedfishPower(rf.accessInfo, id)
+		if err == nil && len(power.PowerControl) > 0 {
+			res.system = id
+			res.consumedWatts = power.PowerControl[0].PowerConsumedWatts
+			res.timestamp = time.Now()
+			rf.systems = append(rf.systems, &res)
+			klog.V(5).Infof("power info: %+v\n", power)
+			if power.PowerControl[0].PowerMetrics.IntervalInMin > intervalInMin {
+				intervalInMin = power.PowerControl[0].PowerMetrics.IntervalInMin
+			}
+		} else {
+			klog.V(5).Infof("failed to get power info: %v\n", err)
+		}
+	}
+
+	// set a timer to check the power info every intervalInMin minutes
+	if intervalInMin > 0 {
+		rf.ticker = time.NewTicker(time.Duration(intervalInMin) * time.Minute)
+		go func() {
+			for {
+				<-rf.ticker.C
+				for _, system := range rf.systems {
+					power, err := getRedfishPower(rf.accessInfo, system.system)
+					if err == nil && len(power.PowerControl) > 0 {
+						// mutex
+						rf.mutex.Lock()
+						klog.V(5).Infof("power info: %+v\n", power)
+						system.consumedWatts = power.PowerControl[0].PowerConsumedWatts
+						system.timestamp = time.Now()
+						rf.mutex.Unlock()
+					} else {
+						klog.V(5).Infof("failed to get power info: %v\n", err)
+					}
+				}
+			}
+		}()
+	}
+	return rf.systems != nil && len(rf.systems) > 0
+}
+
+// GetPower returns the power consumption in Watt
+func (rf *RedFishClient) GetPower() map[string]float64 {
+	if rf.systems != nil {
+		power := make(map[string]float64)
+		for _, system := range rf.systems {
+			rf.mutex.Lock()
+			now := time.Now()
+			// calculate the elapsed time since the last power query in seconds
+			elapsed := now.Sub(system.timestamp).Seconds()
+			klog.V(5).Infof("power info: %+v\n", system)
+			power[system.system] = float64(system.consumedWatts * 1000 * int(elapsed)) // convert to mW
```

Comment:
This line seems to calculate value in Joule (`consumedWatts` x `elapsed`), so is it better to name to `GetEnergyFromRedfish` rather than `GetPower` in func name?

Reply:
It makes sense, @rootfs can you rename this function in the Class level?

Reply:
Never-mind, the Class already defines the right function name: `GetEnergyFromPlatform()`